### ISSUE TITLE
Deprecation rule ignores @DeprecatedSinceKotlin

### ISSUE THREAD

**Issue by osipxd** [detekt/detekt#4084]
## Steps to Reproduce

1. Set Kotlin API version to 1.4 in the build script:
   ```kotlin
   tasks.withType<KotlinCompile>().configureEach {
        kotlinOptions {
            apiVersion = "1.4"
            languageVersion = "1.4"
        }
   }
   
   dependencies {
       implementation(kotlin("stdlib-jdk8", version = "1.5.30"))
   }
   ```
2. Use any function with annotation `@DeprecatedSinceKotlin(warningSince = "1.5")`, for example `String.capitalize()`:
   ```kotlin
   fun foo() {
       println("hello".capitalize())
   }
   ```

## Expected Behavior

There are no reports about deprecation

## Observed Behavior

Detekt reports `capitalize()` extension as Deprecation

## Your Environment

* Version of detekt used: 1.18.1
* Version of Gradle used (if applicable): 7.2
* Operating System and version: macOS 11.2.3

**Comment by BraisGabin** [comment]
Nice catch. For sure this is something that the current rule doesn't handle at all.

**Comment by cortinico** [comment]
We can definitely extend the rule to support the `@DeprecatedSinceKotlin` annotation.

The problem that I see here is that we will most likely not be able to correctly get your `apiLevel` and other custom compiler flags you set.

The reason behind this is that Detekt runs its own compilation with different compiler flags.

This is a feature that could be fully implemented once we move to the Detekt compiler plugin.

**Comment by atulgpt** [comment]
Hi, @BraisGabin / @cortinico  is there any way to write TC against a particular Kotlin language version or specific compiler flag?

**Comment by cortinico** [comment]
> is there any way to write TC against a particular Kotlin language version or specific compiler flag?

Nope that's not possible as of today

**Comment by atulgpt** [comment]
Hi, @3flex since we migrated to worker API, is passing the custom compiler flag feasible now?

**Comment by 3flex** [comment]
No that's not related to adding support for the other compiler flags. There are changes needed in multiple modules to add this support, whether through the compiler plugin or detekt's own compilation environment.

**Comment by atulgpt** [comment]
Hi @cortinico, I found a way to write a TC against a particular language version

```kotlin
@Test
fun `does not report when language settings is below than DeprecatedSinceKotlin version`() {
    val code = """
        @Deprecated("deprecation message")
        @DeprecatedSinceKotlin(warningSince = "1.4")
        fun foo() {
            /* no-op */
        }
        
        fun test() {
            foo()
        }
    """.trimIndent()
    val v = object : CommonCompilerArguments() {}
    v.languageVersion = "1.3"
    env.configuration.setupLanguageVersionSettings(v)
    assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
}

@Test
fun `does report when language settings is above than DeprecatedSinceKotlin version`() {
    val code = """
        @Deprecated("deprecation message")
        @DeprecatedSinceKotlin(warningSince = "1.4")
        fun foo() {
            /* no-op */
        }
        
        fun test() {
            foo()
        }
    """.trimIndent()
    val compilerArgs = object : CommonCompilerArguments() {}
    compilerArgs.languageVersion = "1.6"
    env.configuration.setupLanguageVersionSettings(compilerArgs)
    assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
}

@Test
fun `does report when language settings is equal to DeprecatedSinceKotlin version`() {
    val code = """
        @Deprecated("deprecation message")
        @DeprecatedSinceKotlin(warningSince = "1.4")
        fun foo() {
            /* no-op */
        }
        
        fun test() {
            foo()
        }
    """.trimIndent()
    val compilerArgs = object : CommonCompilerArguments() {}
    compilerArgs.languageVersion = "1.4"
    env.configuration.setupLanguageVersionSettings(compilerArgs)
    assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
}
```

Currently `Deprecation` is passing with the above TCs as well.

**Comment by BraisGabin** [comment]
Yes, the problem is that we don't set the correct language level in the core. And the previous problem is to know which version should we set. It should be doable for sure. But we need to make those two things work.

**Comment by 3flex** [comment]
Once https://github.com/detekt/detekt/pull/7399 is merged it should be fairly easy to add this feature.

You can actually start building now. The API version is available at `compilerResources.languageVersionSettings.apiVersion`. This is currently equal to the language version, but that will change once the linked PR is merged, so it's possible to set language & API versions to different values.

**Comment by segunfamisa** [comment]
Hi there! I'm looking for some more work, and was wondering if this is something still being considered, and if it's open to be worked on.

Thank you.

**Comment by BraisGabin** [comment]
It is. You can go for it :)

**Comment by atulgpt** [comment]
Hi @segunfamisa you can also take the help of some research that I did around this. I have added my comment https://github.com/detekt/detekt/discussions/8530
